src/shared/common/base_data_type.h

```diff
@@ -39,6 +39,12 @@
 #include <map>
 #include <vector>
 
+// Avoid automatic memory allocations
```

Comment:
try to use SYCL_DEVICE_ONLY
as referred from https://gitlab.com/libeigen/eigen/-/blob/master/Eigen/src/Core/util/Macros.h#L851

Reply:
Thank you, this seems to solve the usage of generic pointers and debug compilation. However I had to switch to the git version of Eigen, since their latest published release (3.4.0) does not have those macros. I will later try to debug the 3d case to see if I can find the point in which it dereferences an illegal address.

Reply:
This macro is also in the current version of eigen,
please see it in /vcpkg/installed/x64-linux/include/eigen3/Eigen/src/Core/util/Macros.h

Reply:
The problem with 3.4.0 seems to be that enabling EIGEN_USE_SYCL throws errors at compile-time, which have been fixed in their master branch, however defining SYCL_DEVICE_ONLY works also for 3.4.0

https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/src/Core/util/Macros.h#L568
Judging from this definition, it seems that SYCL_DEVICE_ONLY is a macro defined by Eigen when EIGEN_USE_SYCL is enabled and device code is being compiled.

We could either:
- define SYCL_DEVICE_ONLY with 3.4.0, until a new version is released and then we can switch to EIGEN_USE_SYCL
- define EIGEN_USE_SYCL and switch Eigen's dependency to its git version until a new release is available

However, defining SYCL_DEVICE_ONLY without EIGEN_USE_SYCL will exclude lines of codes tailored for SYCL (e.g. https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/Core#L238).

Reply:
If the interface for the codes tailored for SYCL is the same for those with SYCL_DEVICE_ONLY, I would not mind  the loss of some performance.

Reply:
- https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/Core#L238
- https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/Core#L116

These are some of the parts that would be excluded by not using EIGEN_USE_SYCL, so the entirety of [src/Core/arch/SYCL](https://gitlab.com/libeigen/eigen/-/tree/3.4.0/Eigen/src/Core/arch/SYCL), and some library inclusions, most importantly <CL/sycl.hpp> which we would need to include separately (even though we already include <sycl/sycl.hpp> but Eigen expects the CL namespace).

[src/Core/MathFunctions.h](https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/src/Core/MathFunctions.h) and [src/Core/functors/UnaryFunctors.h](https://gitlab.com/libeigen/eigen/-/blob/3.4.0/Eigen/src/Core/functors/UnaryFunctors.h) are using SYCL_DEVICE_ONLY and that seems to be enough to make it run.

For what concerns performance, the difference seems negligible, at least for those functions used by dambreak.

Reply:
@nR3D so we use SYCL_DEVICE_ONLY now and change it to EIGEN_USE_SYCL after Eigen updated for sycl namespace?